sentinel-extension/sentinel-metric-exporter/src/main/java/com/alibaba/csp/sentinel/metric/exporter/jmx/MetricBeanWriter.java

```diff
@@ -0,0 +1,99 @@
+ * MetricBean in {@link MBeanRegistry} or update the value of MetricBean
+ *
+ * @author chenglu
+ * @date 2021-07-01 20:02
+ * @since 1.8.3
+ */
+public class MetricBeanWriter {
+    
+    private final MBeanRegistry mBeanRegistry = MBeanRegistry.getInstance();
+    
+    private static final String DEFAULT_APP_NAME = "sentinel-application";
+    
+    /**
+     * write the MetricNode value to MetricBean
+     * if the MetricBean is not registered into {@link MBeanRegistry},
+     * it will be created and registered into {@link MBeanRegistry}.
+     * else it will update the value of MetricBean.
+     * Notes. if the MetricNode is null, then {@link MetricBean} will be reset.
+     * @param map metricNode value group by resource
+     * @throws Exception write failed exception
+     */
+    public synchronized void write(Map<String, MetricNode> map) throws Exception {
+        if (map == null || map.isEmpty()) {
+            List<MetricBean> metricNodes = mBeanRegistry.listAllMBeans();
+            if (metricNodes == null || metricNodes.isEmpty()) {
+                return;
+            }
+            for (MetricBean metricNode : metricNodes) {
+                metricNode.reset();
+            }
+            return;
+        }
+        String appName = SentinelConfig.getAppName();
+        if (appName == null) {
+            appName = DEFAULT_APP_NAME;
+        }
+        Set<String> existResource = new HashSet<>();
+        // set or update the new value
+        for (MetricNode metricNode : map.values()) {
+            final String mBeanName = "Sentinel:type=" + appName + ",name=\"" + metricNode.getResource()
+                    +"\",classification=\"" + metricNode.getClassification() +"\"";
+            MetricBean metricBean = mBeanRegistry.findMBean(mBeanName);
+            if (metricBean != null) {
+                metricBean.setValueFromNode(metricNode);
+            } else {
+                metricBean = new MetricBean();
+                metricBean.setValueFromNode(metricNode);
+                mBeanRegistry.register(metricBean, mBeanName);
+                RecordLog.info("[MetricBeanWriter] Registering with JMX as Metric MBean [{}]", mBeanName);
+            }
+            existResource.add(mBeanName);
+        }
+        // reset the old value
+        List<MetricBean> metricBeans = mBeanRegistry.listAllMBeans();
+        if (metricBeans == null || metricBeans.isEmpty()) {
+            existResource.clear();
+            return;
+        }
+        for (MetricBean metricBean : metricBeans) {
+            final String mBeanName = "Sentinel:type=" + appName + ",name=\"" + metricBean.getResource()
+                    +"\",classification=\"" + metricBean.getClassification() +"\"";
+            if (existResource.contains(mBeanName)) {
+                continue;
+            }
+            metricBean.reset();
+        }
+        existResource.clear();
```

Comment:
Currently, we use an extra data structure to reset those elements in the registry that are not included in this round of write.

How about introducing a version(or take the current timestamp as a version) in each write? Then no extra structure is needed and it's also more efficient.

And btw, do they have the opportunity to get removed from the registry if they only appear for a short time?

Reply:
This scene should not occur often, I think we can removed it from registry instead of reset it.

Reply:
but, I still need the `existResouce` to know the MetricBean which need to be removed.

Reply:
Yeah. By the version comparison design I mean you can use a version property to identify which are updated in this round. Example:

```
A{ver: 1}
B{ver: 1}
C{ver: 1}
```

After update
```
A{ver: 2}
B{ver: 2}
C{ver: 1}
```

Then you know C is skipped in the previous round. And it should be removed. By doing this no extra data structure is needed. And you can also use timestamp as the version instead of adding a new one.

Reply:
Thank you so much, I will try to change it as your way.